SofaKernel/modules/SofaCore/src/sofa/core/topology/TopologicalMapping.h

```diff
@@ -175,20 +175,20 @@ class TopologicalMapping : public virtual objectmodel::BaseObject
 
         if (stin == nullptr)
         {
-            msg_error(context) << "Creation of " << className(obj) << " topology mapping failed as input topology model is missing or invalid.";
+            arg->logError("Data attribute 'input' does not point to a valid mesh topology and none can be found in the parent node context.");
             return false;
         }
 
         if (stout == nullptr)
         {
-            msg_error(context) << "Creation of " << className(obj) << " topology mapping failed as output topology model is missing or invalid.";
+            arg->logError("Data attribute 'output' does not point to a valid mesh topology and none can be found in the current node context.");
             return false;
         }
 
-        if (static_cast<BaseObject*>(stin) == static_cast<BaseObject*>(stout))
+        if (dynamic_cast<BaseObject*>(stin) == dynamic_cast<BaseObject*>(stout))
         {
             // we should refuse to create mappings with the same input and output model, which may happen if a State object is missing in the child node
-            msg_error(context) << "Creation of " << className(obj) << " topology mapping failed as the same object \"" << stin->getName() << "\" is linked as input and output.";
+            arg->logError("Both the input mesh and the output mesh points to the same mesh topology ('"+stin->getName()+"').");
             return false;
```

Comment:
Hi Hugo,

The idea is to let  the calling function to detect and print the error. This is why the errors are logged in the object instead of sent via the msg_error API. I think this is not a satisfactory solution because it is an approach to error management that is very invasive. 

I think a much better way of doing that would be to use a frame based message handler (as what was done with the EXPECT_MSG_ and EXPECT_NO_MSG_ for testing tests).
This would allow to write:
```cpp
{ 
         MessageCatcher mc;

         tito();
         canCreate();

         if( mc.hasErrorMessage() )
         {
               msg_error(mc) << "reformat and re-emit the messsages "  < mc.msg();    
         }
}
```
which is actually very close to 
```cpp
try {
        
} catch(Error& m){
               throw Error("reformat and re-emit the messsages "+m.msg()); 
}
```

Reply:
Hi @damienmarchal and @hugtalbot 

The object creation can be a little bit confusing. Here the arg->logError **does not necessary means that an error occurred**. It is there to specify why a given object with a given template and a given context cannot be created.

For example, look at the following scene:
```
<Node>
    <MechanicalObject template="Vec2" />
    <TriangleFEMForcefield />
</Node>
```
And here is what the object creation for the TriangleFEMForcefield could look like:
1. TriangleFEMForcefield\<**Vec2Types**\>
  Does it have a mechanical object with Vec2Types in the current context? 
  **NO**
`  logError("No mo with vec2Types found in the current context")`
2. TriangleFEMForcefield\<**Vec3Types**\>
  Does it have a mechanical object with Vec3Types in the current context? 
  **YES**
  -> Instanciate a new TriangleFEMForcefield\<**Vec3Types**\> 
  -> No error shown since at least one combination of the object/template is valid

Now, let's change the mechanical object by giving him an incompatible template:
```
<Node>
    <MechanicalObject template="Vec2Rigid" />
    <TriangleFEMForcefield />
</Node>
```
And here is what the object creation for the TriangleFEMForcefield could look like:
1. TriangleFEMForcefield\<**Vec2Types**\>
  Does it have a mechanical object with Vec2Types in the current context? 
  **NO**
`  logError("No mechanical object with vec2Types found in the current context")`
2. TriangleFEMForcefield\<**Vec3Types**\>
  Does it have a mechanical object with Vec3Types in the current context? 
  **NO**
`  logError("No mechanical object with vec2Types found in the current context")`
  -> The object factory cannot instantiate a TriangleFEMForcefield object because:
      a. No mechanical object with vec2Types found in the current context
      b. No mechanical object with vec3Types found in the current context

**Here the real error is that no TriangleFEMForcefield can be instantiated** 
The logError are there to specify **what combination of object/template were tried and why they failed**.

Hopefully that clears thing up a little bit